### PR TITLE
Added encode(_:to:) method to FileManager 

### PR DESCRIPTION
**Pull request by AlexeiGitH**
This method allows to save an object that conforms to Encodable protocol to a file at specified URL.
Example:
```Swift
func saveArrayTo(fileURL: URL) {
   let array = [1,2,3,4,5] // array conforms to Encodable
   do {
       try try FileManager.encodeJSON(array, to: fileURL)
   }
   catch {
      print(error)
   }
}
```
## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 4.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 1]
```diff
@@ -56,5 +56,19 @@ public extension FileManager {
         return nil
     }
 
+    /// SwifterSwift: Save Encodable object to a given URL.
+    ///
+    /// - Parameters:
```
Comment (LucianoPAlmeida) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: Missing 'T' parameter on docs

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 2]
```diff
@@ -56,5 +56,19 @@ public extension FileManager {
...
+    ///   - url: URL to save object to.
+    /// - Throws: Throws any errors thrown by JSONEncoder or FileManager.
+    public static func encode<T: Encodable>(_ object: T, to url: URL) throws {
```
Comment (LucianoPAlmeida) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: @AlexeiGitH What do you think about adding a JSONEncoder parameter with default value JSONEncoder() so people can use custom encoders here?

File: Sources/Extensions/Foundation/FileManagerExtensions.swift (hunk 1 above)
Comment (AlexeiGitH) [reply Sources/Extensions/Foundation/FileManagerExtensions.swift]: added

File: Tests/FoundationTests/FileManagerExtensionsTests.swift [hunk 3]
```diff
@@ -80,6 +80,19 @@ final class FileManagerExtensionsTests: XCTestCase {
         } catch {}
     }
 
+    func testEncode() {
+        let point = CGPoint(x: 1, y: 2) //CGPoint conforms to Encodable
+        let tmpDirPath = NSTemporaryDirectory()
```
Comment (guykogus) [inline Tests/FoundationTests/FileManagerExtensionsTests.swift]: See [here](https://nshipster.com/temporary-files/). Interesting update regarding how to correctly use temporary directories.

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 4]
```diff
@@ -56,5 +56,21 @@ public extension FileManager {
...
     }
 
+    /// SwifterSwift: Save Encodable object to a given URL.
+    ///
+    /// - Parameters:
+    ///   - object: Object of type T that conforms to Encodable protocol.
+    ///   - url: URL to save object to.
+    ///   - encoder: JSONEncoder.
+    /// - Throws: Throws any errors thrown by JSONEncoder or FileManager.
+    public static func encode<T: Encodable>(_ object: T, to url: URL, using encoder: JSONEncoder=JSONEncoder()) throws {
```
Comment (guykogus) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: This should not be a static function. Add spacing for the encoder param.
```suggestion
    public func encode<T: Encodable>(_ object: T, to url: URL, using encoder: JSONEncoder = JSONEncoder()) throws {
```

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 5]
```diff
@@ -56,5 +56,21 @@ public extension FileManager {
...
+        do {
```
Comment (guykogus) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: There's no need to use `do-catch` because this function `throws`.

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 6]
```diff
@@ -56,5 +56,21 @@ public extension FileManager {
...
+            let encoder = encoder
+            let data = try encoder.encode(object)
+            FileManager.default.createFile(atPath: url.path, contents: data, attributes: nil)
```
Comment (guykogus) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: When this function is not `static` you can call the local `createFile` function.
```suggestion
            createFile(atPath: url.path, contents: data, attributes: nil)
```

File: Sources/Extensions/Foundation/FileManagerExtensions.swift (hunk 4 above)
Comment (guykogus) [reply Sources/Extensions/Foundation/FileManagerExtensions.swift]: You should add the parameter
```Swift
attributes: [FileAttributeKey: Any]? = nil
```
and pass it to `createFile`

File: Sources/Extensions/Foundation/FileManagerExtensions.swift (hunk 6 above)
Comment (guykogus) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: `createFile` returns a boolean value to indicate if it succeeded or failed. I think it would make sense to return it as the return value of this function so that users will be informed of success/failure.
In my project I have a custom `Error` that I `throw` if the call to `createFile` fails, however adding this isn't possible given the structure of SwifterSwift.

File: Tests/FoundationTests/FileManagerExtensionsTests.swift [hunk 7]
```diff
@@ -80,6 +80,19 @@ final class FileManagerExtensionsTests: XCTestCase {
...
+        let fileURL = URL(fileURLWithPath: tmpDirPath.appending("testEncodeFile"))
+
+        do {
```
Comment (guykogus) [inline Tests/FoundationTests/FileManagerExtensionsTests.swift]: Instead of using `do-catch` here, you should use `XCTAssertNoThrow`s.

File: Tests/FoundationTests/FileManagerExtensionsTests.swift [hunk 8]
```diff
@@ -80,6 +80,19 @@ final class FileManagerExtensionsTests: XCTestCase {
...
+            try FileManager.encode(point, to: fileURL)
+            try FileManager.default.removeItem(at: fileURL) //delete created file
```
Comment (guykogus) [inline Tests/FoundationTests/FileManagerExtensionsTests.swift]: No need for the comment at the end.

File: SwifterSwift.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -333,13 +333,6 @@
 		278CA0911F9A9679004918BD /* NSImageExtensionsTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 278CA0901F9A9679004918BD /* NSImageExtensionsTests.swift */; };
 		42F53FEC2039C5AC0070DC11 /* UIStackViewExtensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = 42F53FEB2039C5AC0070DC11 /* UIStackViewExtensions.swift */; };
 		42F53FF02039C7140070DC11 /* UIStackViewExtensionsTest.swift in Sources */ = {isa = PBXBuildFile; fileRef = 42F53FEF2039C7140070DC11 /* UIStackViewExtensionsTest.swift */; };
-		664CB983217243D200FC87B4 /* DispatchQueueExtensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = 664CB982217243D200FC87B4 /* DispatchQueueExtensions.swift */; };
```
Comment (guykogus) [inline SwifterSwift.xcodeproj/project.pbxproj]: Please revert all these changes.

File: Tests/FoundationTests/FileManagerExtensionsTests.swift (hunk 7 above)
Comment (AlexeiGitH) [reply Tests/FoundationTests/FileManagerExtensionsTests.swift]: I have 3 calls that have try in them (make temporaryDirectoryURL, encode, removeItem).
Should I only use `XCTAssertNoThrow` for `encode` method but inside do{} block?

Comment (guykogus) [reply Tests/FoundationTests/FileManagerExtensionsTests.swift]: Or actually, if you want you can keep it like this, just get rid of the `XCTAssert(true)` and instead create other assertions to check:
1. The new file exists
2. It contains the expected JSON value
3. It no longer exists after calling `removeItem`

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 10]
```diff
@@ -56,5 +56,19 @@ public extension FileManager {
...
+    /// SwifterSwift: Save Encodable object to a given URL.
+    ///
+    /// - Parameters:
+    ///   - object: Object of type T that conforms to Encodable protocol.
+    ///   - url: URL to save object to.
+    ///   - encoder: JSONEncoder.
+    /// - Throws: Throws any errors thrown by JSONEncoder or FileManager.
+    public func encode<T: Encodable>(_ object: T, to url: URL, using encoder: JSONEncoder = JSONEncoder(), attributes: [FileAttributeKey: Any]? = nil) throws -> Bool {
+        let data = try encoder.encode(object)
+        if createFile(atPath: url.path, contents: data, attributes: attributes) {
```
Comment (guykogus) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: Why not just call:
```suggestion
        return createFile(atPath: url.path, contents: data, attributes: attributes)
```
?

Comment (AlexeiGitH) [reply Sources/Extensions/Foundation/FileManagerExtensions.swift]: @guykogus, To perform a second check (Check that file contains the expected JSON value), I am assuming I need to decode the file? (I will add a decode() extension method in another PR so I don't add more than 1 extension in this one)

**Comment by AlexeiGitH** [comment]
@guykogus, I have reverted change for getting tempDir url. 
`let temporaryDirectoryURL = FileManager.default.temporaryDirectory` works for all devices, but the previous implementation with try block fails on iOS.

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,8 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
 # Upcoming release
 
 ### Added
+- **Encodable**
+- Added  `encode(_:to:)` method to save Encodable object to file at provided URL. [#598](https://github.com/SwifterSwift/SwifterSwift/pull/598) by [AlexeiGitH](https://github.com/AlexeiGitH)
```
Comment (guykogus) [inline CHANGELOG.md]: I know I'm being pedantic but... 😛 
```suggestion
- Added `encode(_:to:)` method to save Encodable object to file at provided URL. [#598](https://github.com/SwifterSwift/SwifterSwift/pull/598) by [AlexeiGitH](https://github.com/AlexeiGitH)
```

Comment (AlexeiGitH) [reply CHANGELOG.md]: Fixed. It's so easy to miss because of that `

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 12]
```diff
@@ -56,5 +56,16 @@ public extension FileManager {
         return nil
     }
 
+    /// SwifterSwift: Save Encodable object to a given URL.
+    ///
+    /// - Parameters:
+    ///   - object: Object of type T that conforms to Encodable protocol.
+    ///   - url: URL to save object to.
+    ///   - encoder: JSONEncoder.
```
Comment (LucianoPAlmeida) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: Missing attributes parameter on docs :)

File: CHANGELOG.md [hunk 13]
```diff
@@ -4,6 +4,8 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
 # Upcoming release
 
 ### Added
+- **Encodable**
```
Comment (LucianoPAlmeida) [inline CHANGELOG.md]: Isn't a `FileManager` extension?

File: Sources/Extensions/Foundation/FileManagerExtensions.swift [hunk 14]
```diff
@@ -56,5 +56,16 @@ public extension FileManager {
...
+    /// - Throws: Throws any errors thrown by JSONEncoder or FileManager.
+    public func encode<T: Encodable>(_ object: T, to url: URL, using encoder: JSONEncoder = JSONEncoder(), attributes: [FileAttributeKey: Any]? = nil) throws -> Bool {
```
Comment (guykogus) [inline Sources/Extensions/Foundation/FileManagerExtensions.swift]: I think you should switch `encoder` and `url` parameters. It reads more naturally because of the `attributes` parameter that it would paired with `url`.
Also, you need to add `attributes` documentation.

File: CHANGELOG.md (hunk 13 above)
Comment (AlexeiGitH) [reply CHANGELOG.md]: @LucianoPAlmeida Fixed

File: Tests/FoundationTests/FileManagerExtensionsTests.swift [hunk 15]
```diff
@@ -80,6 +80,28 @@ final class FileManagerExtensionsTests: XCTestCase {
...
     }
 
+    func testEncode() {
+        let point = CGPoint(x: 1, y: 2) //CGPoint conforms to Encodable
+        do {
+            let temporaryDirectoryURL = FileManager.default.temporaryDirectory
+            let temporaryFilename = ProcessInfo().globallyUniqueString
+            let temporaryFileURL = temporaryDirectoryURL.appendingPathComponent(temporaryFilename)
+
+            XCTAssertNoThrow(try FileManager.default.encode(point, to: temporaryFileURL))
```
Comment (guykogus) [inline Tests/FoundationTests/FileManagerExtensionsTests.swift]: No need for the assertion here since you're `catch`-ing. It won't make sense to run the rest of the test anyway if this fails, so better that it be caught.

File: Tests/FoundationTests/FileManagerExtensionsTests.swift [hunk 16]
```diff
@@ -80,6 +80,28 @@ final class FileManagerExtensionsTests: XCTestCase {
...
+            XCTAssert(FileManager.default.fileExists(atPath: temporaryFileURL.path))
+
+            let data = FileManager.default.contents(atPath: temporaryFileURL.path)
+            XCTAssertNotNil(data)
+
+            let decoder = JSONDecoder()
+            XCTAssertNoThrow(try decoder.decode(CGPoint.self, from: data!))
```
Comment (guykogus) [inline Tests/FoundationTests/FileManagerExtensionsTests.swift]: Again, no assertion. Get the value and check that it matches the original.

File: Sources/Extensions/Foundation/DateExtensions.swift [hunk 17]
```diff
@@ -455,7 +455,9 @@ public extension Date {
     ///     let yesterday = date.yesterday // "Oct 2, 2018, 10:57:11"
     ///
     public var yesterday: Date {
-        return addingTimeInterval(-86400.0)
+        let cal = Calendar.current
```
Comment (guykogus) [inline Sources/Extensions/Foundation/DateExtensions.swift]: Please do this in a different PR

Comment (guykogus) [reply Sources/Extensions/Foundation/DateExtensions.swift]: @AlexeiGitH please remove these changes and fix the conflicts so that we can finally merge this in 😅

File: Tests/FoundationTests/FileManagerExtensionsTests.swift [hunk 18]
```diff
@@ -80,6 +80,30 @@ final class FileManagerExtensionsTests: XCTestCase {
...
+
+            let data = FileManager.default.contents(atPath: temporaryFileURL.path)
+            XCTAssertNotNil(data)
+
+            let decoder = JSONDecoder()
+            let decodedPoint = try decoder.decode(CGPoint.self, from: data!)
+            XCTAssertTrue(point == decodedPoint)
+
+            try FileManager.default.removeItem(at: temporaryFileURL)
+            XCTAssertFalse(FileManager.default.fileExists(atPath: temporaryFileURL.path))
```
Comment (guykogus) [inline Tests/FoundationTests/FileManagerExtensionsTests.swift]: You don't really need this check, you can actually change the line above to:
```Swift
XCTAssertTrue(try FileManager.default.removeItem(at: temporaryFileURL))
```
Even that you don't really need, because if the removal fails it will throw an exception, which you will catch.
Sorry that I didn't catch this in the last review... This one seems to be going on forever 😅

**Comment by guykogus** [comment]
@omaralbeik @LucianoPAlmeida can you please check?

**Comment by LucianoPAlmeida** [comment]
Sorry, I just notice that we missed that one 😅, just cleaning up those old PR's. 
@AlexeiGitH Closing this one due inactivity, feel free to reopen :))
